File: solutions/csharp/pythagorean-triplet/1/PythagoreanTriplet.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class PythagoreanTriplet
{
    public static IEnumerable<(int a, int b, int c)> TripletsWithSum(int sum) =>
        GenerateTriplets(sum)
            .Where(triplet => triplet.a + triplet.b + triplet.c == sum);

    private static IEnumerable<(int a, int b, int c)> GenerateTriplets(int sum) =>
        GenerateTriplets(1, sum, new List<(int a, int b, int c)>());

    private static IEnumerable<(int a, int b, int c)> GenerateTriplets(int min, int max, List<(int a, int b, int c)> accumulator)
    {
        if (min > max)
        {
            return accumulator;
        }

        List<(int a, int b, int c)> triplets = Enumerable.Range(min, max)
            .Select(value => CalculateDistance(min, value))
            .Where(triplet => IsWholeNumber(triplet.c))
            .Select(triplet => (triplet.a, triplet.b, Convert.ToInt32(triplet.c)))
            .ToList();

        accumulator.AddRange(triplets);
        return GenerateTriplets(min + 1, max, accumulator);
    }

    private static bool IsWholeNumber(decimal value) =>
        value - Math.Round(value) == 0;

    private static (int a, int b, decimal c) CalculateDistance(int a, int b) =>
        (a, b, SquareRoot(Square(a) + Square(b)));

    private static int Square(int value) =>
        value * value;

    private static decimal SquareRoot(long value)
    {
        decimal sqrt_x = (decimal)Math.Sqrt(value);
        for (int i = 0; i < 10; ++i)
        {
            sqrt_x = 0.5m * (sqrt_x + value / sqrt_x);
        }
        return sqrt_x;
    }
}

File: solutions/scala/pythagorean-triplet/1/src/main/scala/PythagoreanTriplet.scala
object PythagoreanTriplet:
    private type Triplet = (Int, Int, Int)

    def isPythagorean(triplet: Triplet): Boolean =
        val (a, b, c) = triplet
        a * a + b * b == c * c

    def pythagoreanTriplets(
                               minimumFactor: Int,
                               maximumFactor: Int
                           ): Seq[Triplet] =
        for
            a <- minimumFactor until maximumFactor - 1
            b <- (a + 1) until maximumFactor
            c <- (b + 1) to maximumFactor
            triplet = (a, b, c)
            if isPythagorean(triplet)
        yield triplet

    def pythagoreanTripletsSum(sum: Int): Seq[Triplet] =
        for
            a <- 1 until sum / 3
            b <- (a + 1) until sum / 2
            c = sum - a - b
            triplet = (a, b, c)
            if b < c
            if isPythagorean(triplet)
        yield triplet